node/core/dispute-coordinator/src/real/participation/queues.rs

```diff
@@ -0,0 +1,188 @@
+	/// a quick resolution of disputes, even in the best effort queue.
+	best_effort: HashMap<CandidateHash, BestEffortEntry>,
+
+	/// Priority queue.
+	///
+	/// In the priority queue, we have a strict ordering of candidates and participation will
+	/// happen in that order.
+	priority: BTreeMap<CandidateComparator, ParticipationRequest>,
+}
+
+/// A dispute participation request that can be queued.
+pub struct ParticipationRequest {
+	candidate_hash: CandidateHash,
+	candidate_receipt: CandidateReceipt,
+	session: SessionIndex,
+	n_validators: usize,
+}
+
+/// Entry for the best effort queue.
+struct BestEffortEntry {
+	req: ParticipationRequest,
+	/// How often was the above request added to the queue.
+	added_count: BestEffortCount,
+}
+
+/// What can go wrong when queuing a request.
+#[derive(Debug, Error)]
+pub enum Error {
+	#[error("Request could not be queued, because best effort queue was already full.")]
+	BestEffortFull,
+	#[error("Request could not be queued, because priority queue was already full.")]
+	PriorityFull,
+}
+
+impl ParticipationRequest {
+	/// Create a new `ParticipationRequest` to be queued.
+	pub fn new(
+		candidate_receipt: CandidateReceipt,
+		session: SessionIndex,
+		n_validators: usize,
+	) -> Self {
+		Self { candidate_hash: candidate_receipt.hash(), candidate_receipt, session, n_validators }
+	}
+
+	pub fn candidate_receipt(&'_ self) -> &'_ CandidateReceipt {
+		&self.candidate_receipt
+	}
+	pub fn candidate_hash(&'_ self) -> &'_ CandidateHash {
+		&self.candidate_hash
+	}
+	pub fn session(&self) -> SessionIndex {
+		self.session
+	}
+	pub fn n_validators(&self) -> usize {
+		self.n_validators
+	}
+	pub fn into_candidate_info(self) -> (CandidateHash, CandidateReceipt) {
+		let Self { candidate_hash, candidate_receipt, .. } = self;
+		(candidate_hash, candidate_receipt)
+	}
+}
+
+impl Queues {
+	/// Create new `Queues`.
+	pub fn new() -> Self {
+		Self { best_effort: HashMap::new(), priority: BTreeMap::new() }
+	}
+
+	/// Will put message in queue, either priority or best effort depending on whether a
+	/// `CandidateComparator` was provided or not.
+	///
+	/// If the message was already previously present on best effort, it will be moved to priority
+	/// if a `CandidateComparator` has been passed now, otherwise the `added_count` on the best
+	/// effort queue will be bumped.
+	///
+	/// Returns error in case a queue was found full already.
+	pub fn queue(
+		&mut self,
+		comparator: Option<CandidateComparator>,
+		req: ParticipationRequest,
+	) -> Result<(), Error> {
+		if let Some(comparator) = comparator {
+			if self.priority.len() >= PRIORITY_QUEUE_SIZE {
+				return Err(Error::PriorityFull)
+			}
+			// Remove any best effort entry:
+			self.best_effort.remove(&req.candidate_hash);
+			self.priority.insert(comparator, req);
+		} else {
+			if self.best_effort.len() >= BEST_EFFORT_QUEUE_SIZE {
+				return Err(Error::BestEffortFull)
+			}
+			self.best_effort
+				.entry(req.candidate_hash)
```

Comment:
`queue(None, req)` can lead to something being present in both best effort and participation queues.

Reply:
Good catch!